File: android/a-kart/app/src/main/java/com/frogdesign/akart/DiscoveryActivity.java
package com.frogdesign.akart;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import com.frogdesign.arsdk.Discovery;
import com.frogdesign.akart.databinding.DiscoveryListItemBinding;
import com.parrot.arsdk.ardiscovery.ARDiscoveryDeviceService;

import java.util.ArrayList;
import java.util.List;

import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subjects.PublishSubject;

public class DiscoveryActivity extends AppCompatActivity {
    private static final String TAG = DiscoveryActivity.class.getSimpleName();
    private RecyclerView list;
    private Button search;
    private DiscoveryAdapter adapter;
    private LinearLayoutManager llm;
    private Discovery discovery;
    private Subscription sub;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        DataBindingUtil.setContentView(this, R.layout.discovery_activity);
        discovery = new Discovery(getBaseContext());
        list = (RecyclerView) findViewById(android.R.id.list);
        list.setHasFixedSize(true);
        llm = new LinearLayoutManager(this);
        list.setLayoutManager(llm);
        adapter = new DiscoveryAdapter(this);
        list.setAdapter(adapter);

        adapter.clicksOn().subscribe(dev -> {
            Log.i("CLICK", "on " + dev.getName());
            Intent i = new Intent(getBaseContext(), PlayActivity.class);
            i.putExtra(PlayActivity.EXTRA_DEVICE, dev);
            startActivity(i);
        });

        search = (Button) findViewById(R.id.search);
        search.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startDiscovery();
            }
        });
    }

    @Override
    protected void onStart() {
        super.onStart();
        startDiscovery();
    }

    private void startDiscovery() {
        if (sub != null && !sub.isUnsubscribed()) {
            sub.unsubscribe();
            sub = null;

        }
        sub = discovery.discoverer()
//        ;
//        TestUtils.constantDiscoverer()
                .subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(deviceList -> {
            Log.i(TAG, "--> SERVICES:");
            // Do what you want with the device list
            for (ARDiscoveryDeviceService service : deviceList) {
                Log.i(TAG, "The service " + service);
            }
            Log.i(TAG, "<-- SERVICES.");
            adapter.update(deviceList);
            //Controller ctrl = new Controller(getBaseContext(), deviceList.get(0));
            //ctrl.start();
            //ctrl.mediaStreamer().observeOn(AndroidSchedulers.mainThread()).subscribe(bmpConsumer);
        });
    }

    public static class DiscoveryAdapter extends RecyclerView.Adapter<DiscoveryListItemHolder>
            implements View.OnClickListener {

        private final LayoutInflater layoutInflater;
        private final PublishSubject<ARDiscoveryDeviceService> subject = PublishSubject.create();
        private final ArrayList<ARDiscoveryDeviceService> devices = new ArrayList<>();

        private DiscoveryAdapter(Context ctx) {
            layoutInflater = LayoutInflater.from(ctx);
            setHasStableIds(true);
        }

        private void update(List<ARDiscoveryDeviceService> newDevices) {
            devices.clear();
            devices.addAll(newDevices);
            notifyDataSetChanged();
        }

        @Override
        public long getItemId(int position) {
            return devices.get(position).getName().hashCode();
        }

        @Override
        public DiscoveryListItemHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            DiscoveryListItemBinding binding =
                    DiscoveryListItemBinding.inflate(layoutInflater, parent, false);
            binding.setHandler(this);
            return new DiscoveryListItemHolder(binding);
        }

        @Override
        public void onBindViewHolder(DiscoveryListItemHolder holder, int position) {
            holder.bind(devices.get(position));
        }

        @Override
        public int getItemCount() {
            return devices.size();
        }

        public Observable<ARDiscoveryDeviceService> clicksOn() {
            return subject;
        }

        @Override
        public void onClick(View v) {
            ARDiscoveryDeviceService device = (ARDiscoveryDeviceService) v.getTag();
            subject.onNext(device);
        }
    }

    public static class DiscoveryListItemHolder extends RecyclerView.ViewHolder {
        private DiscoveryListItemBinding binding;

        private DiscoveryListItemHolder(DiscoveryListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        private void bind(ARDiscoveryDeviceService device) {
            itemView.setTag(device);
            binding.setDevice(device);
            binding.executePendingBindings();
        }
    }
}

File: android/a-kart/app/src/main/kotlin/com/frogdesign/akart/DiscoveryActivity.kt
package com.frogdesign.akart

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.AdapterDataObserver
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import butterknife.bindView
import com.frogdesign.akart.model.Car
import com.frogdesign.akart.model.Cars
import com.frogdesign.akart.view.CheckableImageView
import com.parrot.arsdk.ardiscovery.ARDiscoveryDeviceService
import org.lucasr.twowayview.ItemSelectionSupport
import rx.Observable
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subjects.PublishSubject
import timber.log.Timber
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper
import java.util.*
import kotlin.properties.Delegates

class DiscoveryActivity : AppCompatActivity() {
    private val list: RecyclerView by bindView(android.R.id.list)
    private val swiperefresh: SwipeRefreshLayout by bindView(R.id.swiperefresh)
    private val search: Button by bindView(R.id.search)
    private val prompt: TextView by bindView(R.id.prompt)

    private var adapter: DiscoveryAdapter by Delegates.notNull()
    private var discovery: Discovery by Delegates.notNull()
    private var selections: ItemSelectionSupport by Delegates.notNull()
    private var sub: Subscription? = null

    private val handler = Handler()
    private val stopRefresh = Runnable { swiperefresh.isRefreshing = false }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.discovery_activity)
        discovery = Discovery(baseContext)
        selections = ItemSelectionSupport.addTo(list)
        adapter = DiscoveryAdapter(this)

        list.layoutManager = LinearLayoutManager(this)
        list.setHasFixedSize(true)
        selections.choiceMode = ItemSelectionSupport.ChoiceMode.MULTIPLE
        list.adapter = adapter

        adapter.registerAdapterDataObserver(object : AdapterDataObserver() {
            override fun onChanged() {
                updateButton()
            }
        })

        search.setOnClickListener {
            val i = Intent(baseContext, PlayActivity::class.java)
            val checks = selections.checkedItemPositions
            if (checks.size() != 1)
                throw IllegalArgumentException("Should be 1 checked!")

            val checked = checks.keyAt(0)
            val car = adapter.getItemAt(checked)
            i.putExtra(PlayActivity.EXTRA_DEVICE, car.associatedDevice)
            startActivity(i)
        }
        swiperefresh.setOnRefreshListener {
            startDiscovery()
            selections.clearChoices()
            adapter.update(emptyList())
        }

        swiperefresh.setColorSchemeResources(R.color.militaryGreen, R.color.militaryGreen, R.color.militaryGreenDark)
        swiperefresh.setProgressBackgroundColorSchemeColor(Color.TRANSPARENT)
    }

    override fun attachBaseContext(newBase : Context) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
    }

    override fun onStop() {
        super.onStop()
        ensureUnsubscribed()
        discovery.unbind()
        stopRefresh.run()
        handler.removeCallbacks(stopRefresh)
    }

    private fun ensureUnsubscribed() {
        if (sub != null && !sub!!.isUnsubscribed) {
            sub!!.unsubscribe()
            sub = null
        }
    }

    private fun startDiscovery() {
        ensureUnsubscribed()
        sub = discovery.discoverer().subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribe({ deviceList ->
            Timber.tag(TAG).i("--> SERVICES:")
            // Do what you want with the device list
            for (service in deviceList) {
                Timber.tag(TAG).i( "The service " + service)
            }
            Timber.tag(TAG).i( "<-- SERVICES.")

            @Suppress("UNCHECKED_CAST")
//            val dl = listOf(ARDiscoveryDeviceService("gargamella", null, 0),
//                                ARDiscoveryDeviceService("taxiguerrilla", null, 1))
            adapter.update(deviceList.map { dev -> Cars.retrieveRelatedTo(dev) }.filter { it -> it != null } as List<Car>)

        })

        swiperefresh.isRefreshing = true
        handler.removeCallbacks(stopRefresh)
        handler.postDelayed(stopRefresh, 3000);
    }

    override fun onResume() {
        super.onResume()
        updateButton()
    }

    private fun updateButton() {
        val size = selections.checkedItemPositions.size()
        search.isEnabled = size > 0

        val elems = adapter.itemCount
        prompt.visibility = if (elems > 0) View.INVISIBLE else View.VISIBLE
    }

    inner class DiscoveryAdapter constructor(ctx: Context) : RecyclerView.Adapter<DiscoveryListItemHolder>(), View.OnClickListener {

        private val layoutInflater: LayoutInflater
        private val devices = ArrayList<Car>()

        init {
            layoutInflater = LayoutInflater.from(ctx)
            setHasStableIds(true)
        }

        fun update(newDevices: List<Car>) {
            devices.clear()
            devices.addAll(newDevices)
            notifyDataSetChanged()
        }

        override fun getItemId(position: Int): Long {
            return devices[position].id.hashCode().toLong()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DiscoveryListItemHolder {
            val v = layoutInflater.inflate(R.layout.discovery_list_item, parent, false)
            return DiscoveryListItemHolder(v, this)
        }

        override fun onBindViewHolder(holder: DiscoveryListItemHolder, position: Int) {
            holder.bind(devices[position])
        }

        override fun getItemCount(): Int {
            return devices.size
        }

        override fun onClick(v: View) {
            val indexClicked = list.getChildAdapterPosition(v)
            val currentlyChecked = selections.isItemChecked(indexClicked)
            selections.clearChoices()
            Timber.i("Clicked on %s, currently %s", indexClicked, true)
            if (currentlyChecked) selections.setItemChecked(indexClicked, currentlyChecked)
            notifyDataSetChanged()
        }

        fun getItemAt(checked: Int): Car = devices[checked]
    }

    inner class DiscoveryListItemHolder constructor(v: View, private val on: View.OnClickListener) : RecyclerView.ViewHolder(v) {
        private val selectable: CheckableImageView
        private val avatar: ImageView
        private val text1: TextView

        init {
            selectable = v.findViewById(R.id.selectable) as CheckableImageView
            avatar = v.findViewById(R.id.avatar) as ImageView
            text1 = v.findViewById(android.R.id.text1) as TextView
        }

        fun bind(device: Car) {
            itemView.tag = device
            text1.text = device.id
            itemView.setOnClickListener(on)
            Timber.i("BIND index %s on %s", adapterPosition, selections.isItemChecked(adapterPosition))
            selectable.isChecked = selections.isItemChecked(adapterPosition) ?: false
        }
    }

    companion object {
        private val TAG = DiscoveryActivity::class.java.simpleName
    }
}
